src/docs/binding-classes.md

```diff
@@ -0,0 +1,397 @@
+
+The allocating initializer takes the type metadata for the class in the `self` register. We can call this directly, however under enable-library-evolution, the swift compiler generates a dispatch thunk for this method and we should probably call that.
+
+Swift divides initializers into two categories: designated and convenience. Designated initializers are initializers which fully initialize a class. If a subclass inherits from base class, it must implement all designated initializers and call the parent class. A convenience intializer will have a different signature than all desinated initializers and **must** call a designated initializer.
+
+The generic programming model in Swift does specialization at runtime. It does this by creating the type metadata for the specialized type
+using the Metadata Accessor function. The Metadata Accessor function has one type metadata argument for each specialized type in the generic class. It is **very important** that the binding code calls the Metadata Accessor rather than trying to synthesize the type metadata object. This is because from Swift's point of view, every type metadata object is a singleton and the runtime will cache generic specializations. 
+
+Swift initializers can fail - this is something that doesn't have an exact analog in C#. There are two ways to handle this:
+- make the corresponding C# constructor throw
+- write a factory method that returns `BoundType?`
+
+## ABI differences
+- Instance methods get called with the `self` register pointing to the instance.
+- Class methods get called with the `self` register pointing to the type metadata.
+- Init methods get called with the type metadata in the first argument.
+- Methods that throw return the error in the error register.
+- Virtual methods in swift get called via a vtable. Since the ordering of the vtable is undefined, methods can't be called using the vtable from C#. In addition, the vtable is consider fragile and may change between versions of any given library. Fortunately, with enable-library-evolution set, the compiler will inject a dispatch thunk for each virtual method.
+
+
+## Runtime Differences
+Swift classes are reference counted rather than garbage collected. This can create some interesting conditions when an object from Swift surfaced in C# gets garbage collected, but the object may still be live in Swift, so it's necessary to manage this. The way this is handled in Binding Tools for Swift is to add it to an object registry which is a `Dictionary&lt;NativeHandle, GCHandle&gt;` The native handle is the Swift object instance and the GC Handle is made from the corresponding C# instance. When the pair is added to the registry, the registry also takes a weak reference to the Swift
```

Comment:
```suggestion
Swift classes are reference counted rather than garbage collected. This can create some interesting conditions when an object from Swift surfaced in C# gets garbage collected, but the object may still be live in Swift, so it's necessary to manage this. The way this is handled in Binding Tools for Swift is to add it to an object registry which is a `Dictionary<NativeHandle, GCHandle>` The native handle is the Swift object instance and the GC Handle is made from the corresponding C# instance. When the pair is added to the registry, the registry also takes a weak reference to the Swift
```